### PR TITLE
Add support for DPS provisioning flow for pnp samples

### PR DESCRIPTION
**Pull request by abhipsaMisra**
This PR currently has the DPS flow for the Thermostat sample. Once I get a sign-off on the approach, I'll implement the same for the TemperatureController sample as well.

Passing in env var via command line -> defaults to env variable:
image

### CODE REVIEW

File: iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs [hunk 1]
```diff
@@ -27,7 +28,21 @@ public class Program
         // DTDL interface used: https://github.com/Azure/opendigitaltwins-dtdl/blob/master/DTDL/v2/samples/Thermostat.json
         private const string ModelId = "dtmi:com:example:Thermostat;1";
 
+        // This environment variable indicates if DPS or IoT Hub connection string will be used to provision the device.
+        // Expected values: (case-insensitive)
```
Comment (bikamani) [inline iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs:32]: shouldn't we take care of casing?

File: iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs [hunk 2]
```diff
@@ -7,10 +7,11 @@
 using System.Text;
 using System.Threading.Tasks;
 using Microsoft.Azure.Devices.Client;
+using Microsoft.Azure.Devices.Provisioning.Client;
```
Comment (bikamani) [inline iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs:10]: Are you planning to do temperature controller sample changes in a separate PR?

File: iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs (hunk 1 above)
Comment (abhipsaMisra) [reply iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs:32]: no, it doesn't matter; this is only a switch to indicate which flow is to be demonstrated.

File: iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs (hunk 2 above)
Comment (abhipsaMisra) [reply iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs:10]: No, i'll it to this PR, once I get a green from the certification flow.

File: iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs (hunk 1 above)
Comment (bikamani) [reply iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs:32]: you are already doing it actually.

File: iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs (hunk 2 above)
Comment (bikamani) [reply iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs:10]: 👍

File: iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs [hunk 3]
```diff
@@ -27,7 +28,21 @@ public class Program
         // DTDL interface used: https://github.com/Azure/opendigitaltwins-dtdl/blob/master/DTDL/v2/samples/Thermostat.json
         private const string ModelId = "dtmi:com:example:Thermostat;1";
 
+        // This environment variable indicates if DPS or IoT Hub connection string will be used to provision the device.
+        // Expected values: (case-insensitive)
+        // "DPS" - The sample will use DPS to provision the device.
+        // "connectionString" - The sample will use IoT Hub connection string to provision the device.
+        private static readonly string s_deviceSecurityType = Environment.GetEnvironmentVariable("IOTHUB_DEVICE_SECURITY_TYPE");
```
Comment (abhipsaMisra) [inline iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs:35]: Since it is a requirement to have a similar usage experience across all sdks, all languages are using environment variables, and not language specific command line parsers (even though they are definitely more convenient).

File: iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs [hunk 4]
```diff
@@ -27,7 +28,21 @@ public class Program
...
+
+        // Required if IOTHUB_DEVICE_SECURITY_TYPE is set to "connectionString".
         private static readonly string s_deviceConnectionString = Environment.GetEnvironmentVariable("IOTHUB_DEVICE_CONNECTION_STRING");
```
Comment (drwill-ms) [inline iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs:38]: Any interest in using the CommandLineParser nuget, like I changed up recently in our samples repo? It isn't that big of a change. https://github.com/Azure-Samples/azure-iot-samples-csharp/commit/6097ef17fe76875c40241136cd45439ca187f21c

File: iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs [hunk 5]
```diff
@@ -66,8 +81,36 @@ private static async Task RunSampleAsync()
...
+                        var authMethod = new DeviceAuthenticationWithRegistrySymmetricKey(dpsRegistrationResult.DeviceId, s_deviceSymmetricKey);
+                        InitializeDeviceClientAsync(dpsRegistrationResult.AssignedHub, authMethod);
+                        break;
+                    }
+                    throw new ArgumentException("Required environment variables are not set for DPS flow, please recheck your environment.");
+
+                case "connectionstring":
+                    if (ValidateArgsForIotHubFlow())
+                    {
+                        InitializeDeviceClientAsync(s_deviceConnectionString);
```
Comment (drwill-ms) [inline iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs:105]: no await?

File: iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs (hunk 4 above)
Comment (abhipsaMisra) [reply iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs:38]: Yes, I did want to do that, but maintaining similar usage experience across all languages was important here. So the decision was taken for all languages to use environment variables.

File: iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs [hunk 6]
```diff
@@ -66,8 +81,36 @@ private static async Task RunSampleAsync()
...
             s_logger.LogDebug($"Initialize the device client.");
-            InitializeDeviceClientAsync();
+            switch (s_deviceSecurityType.ToLowerInvariant())
+            {
+                case "dps":
+                    if (ValidateArgsForDpsFlow())
+                    {
+                        DeviceRegistrationResult dpsRegistrationResult = await ProvisionDeviceAsync();
+                        var authMethod = new DeviceAuthenticationWithRegistrySymmetricKey(dpsRegistrationResult.DeviceId, s_deviceSymmetricKey);
+                        InitializeDeviceClientAsync(dpsRegistrationResult.AssignedHub, authMethod);
```
Comment (drwill-ms) [inline iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs:97]: no await?

File: iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs (hunk 5 above)
Comment (abhipsaMisra) [reply iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs:105]: That is bad naming, this method doesn't do anything async; I'll fix this.

File: iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs (hunk 6 above)
Comment (abhipsaMisra) [reply iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs:97]: same here as well, shouldn't be called async. I'll fix this

File: iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs [hunk 7]
```diff
@@ -66,8 +81,36 @@ private static async Task RunSampleAsync()
...
+                        break;
+                    }
+                    throw new ArgumentException("Required environment variables are not set for DPS flow, please recheck your environment.");
```
Comment (abhipsaMisra) [inline iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs:100]: I am thinking about changing this to log error and then exit (as opposed to throwing an exception). Thoughts?

File: iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs (hunk 3 above)
Comment (drwill-ms) [reply iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs:35]: Who set that requirement?

File: iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs (hunk 4 above)
Comment (drwill-ms) [reply iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs:38]: We have argument handling all mixed in with the rest of the sample. This is bad form and makes it hard to read. Can you do that checking at program init?

Comment (drwill-ms) [reply iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs:38]: Here's a solution to have both https://github.com/commandlineparser/commandline/issues/677

Comment (drwill-ms) [reply iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs:38]: Perhaps not, it seems.

File: iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs (hunk 3 above)
Comment (abhipsaMisra) [reply iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs:35]: This was set when we first started writing the samples this summer. The idea was that, if someone where to give the python sample a try, and then decide to try the .NET sample, the on-boarding time should be min, and experience should be similar.
We know that java and .NET have similar experience with these command line parsers, I'll sync up with scripting team to see if they have a similar option. If not, we might have to stick to env var for now.

File: iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs (hunk 4 above)
Comment (abhipsaMisra) [reply iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs:38]: I agree, it does make the sample cluttered. I'll decouple the args validation from the main sample.

**Review by drwill-ms — APPROVED** [review]

File: iothub/device/samples/PnpDeviceSamples/Thermostat/Parameter.cs [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+{
+    /// <summary>
+    /// Parameters for the application supplied via command line arguments. It default to environment variables..
+    /// </summary>
+    internal class Parameters
+    {
+        [Option(
+            's',
+            "DeviceProvisioningType",
+            HelpText = "(Required) The provisioning type that will be used for provisioning the device for the sample. Possible values include dps, hubConnectionString (case-insensitive)." +
```
Comment (abhipsaMisra) [inline iothub/device/samples/PnpDeviceSamples/Thermostat/Parameter.cs:17]: Cannot set the option as "required", since then we cannot dynamically resolve the value if it needs to be sent via environment variables. This is a trade-off in the sample supporting both env var and command line args.

File: iothub/device/samples/PnpDeviceSamples/Thermostat/Parameter.cs [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+            if (string.IsNullOrWhiteSpace(DeviceProvisioningType))
+            {
+                throw new ArgumentNullException(nameof(DeviceProvisioningType), "Device provisioning type needs to be specified, please set the environment variable \"IOTHUB_DEVICE_SECURITY_TYPE\"" +
+                    "or pass in \"-s | --DeviceProvisioningType\" through command line.");
+            }
+
+            return (DeviceProvisioningType.ToLowerInvariant()) switch
+            {
+                "dps" => !string.IsNullOrWhiteSpace(DpsEndpoint)
+                                       && !string.IsNullOrWhiteSpace(DpsIdScope)
```
Comment (drwill-ms) [inline iothub/device/samples/PnpDeviceSamples/Thermostat/Parameter.cs:67]: excessive indenting

File: iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs [hunk 10]
```diff
@@ -2,46 +2,61 @@
...
+                    parameters = parsedParams;
+                })
+                .WithNotParsed(errors =>
+                {
+                    Environment.Exit(1);
+                });
+
+            if (!parameters.Validate())
+            {
+                s_logger.LogError("Required parameters are not set, please recheck required variables by running \"dotnet run -- --help\"");
```
Comment (drwill-ms) [inline iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs:39]: period after `are not set`, as the next bit is its own sentence.

Comment (drwill-ms) [inline iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs:39]: If someone has the compiled sample, and not the csproj, `dotnet run` doesn't help them.

File: iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs [hunk 11]
```diff
@@ -52,179 +67,82 @@ public static async Task Main(string[] _)
...
-                        // Generate a random value between 5.0°C and 45.0°C for the current temperature reading.
-                        s_temperature = Math.Round(s_random.NextDouble() * 40.0 + 5.0, 1);
-                        temperatureReset = false;
-                    }
-
-                    await SendTemperatureAsync();
-                    await Task.Delay(5 * 1000);
-                }
-            });
+                JsonData = $"{{ \"modelId\": \"{ModelId}\" }}",
```
Comment (drwill-ms) [inline iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs:107]: When do we plan on addressing the needing to know this format and hard code?

File: iothub/device/samples/PnpDeviceSamples/Thermostat/ThermostatSample.cs [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+            }
+        }
+
+        // Send temperature update over telemetry.
+        private async Task SendTemperatureTelemetryAsync()
+        {
+            string telemetryName = "temperature";
+
+            string telemetryPayload = $"{{ \"{telemetryName}\": {_temperature} }}";
+            var message = new Message(Encoding.UTF8.GetBytes(telemetryPayload))
```
Comment (drwill-ms) [inline iothub/device/samples/PnpDeviceSamples/Thermostat/ThermostatSample.cs:176]: using?

File: iothub/device/samples/PnpDeviceSamples/Thermostat/Parameter.cs [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+namespace Thermostat
+{
+    /// <summary>
+    /// Parameters for the application supplied via command line arguments. It default to environment variables..
+    /// </summary>
+    internal class Parameters
+    {
+        [Option(
+            's',
+            "DeviceProvisioningType",
```
Comment (drwill-ms) [inline iothub/device/samples/PnpDeviceSamples/Thermostat/Parameter.cs:16]: DpsSecurityType?

**Review by drwill-ms — APPROVED** [review]

File: iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs (hunk 11 above)
Comment (abhipsaMisra) [reply iothub/device/samples/PnpDeviceSamples/Thermostat/Program.cs:107]: I'll be adding this into a pnp helper that will go into our dps assembly. This helper will be referenced in the `TemperatureController` sample.

File: iothub/device/samples/PnpDeviceSamples/Thermostat/Parameter.cs (hunk 13 above)
Comment (abhipsaMisra) [reply iothub/device/samples/PnpDeviceSamples/Thermostat/Parameter.cs:16]: Updated to DeviceSecurityType -> in line with the env var name.
